### DISCUSSION TITLE
Optional route parameter

### DISCUSSION THREAD

**Discussion by machour** [remix-run/remix#1261] [Q&A] [answered]
I looked through the documentation, but didn't find how to handle optional routes parameters.

Let's say I have a paginated set of results, I would like to be able to have these urls:

`results/` : first page of results (which is also `results/1`)
`results/2` : second page of results
...

The quickest solution I came up with for now, is to have the following routes:
`results/$page.tsx`

```ts
export const loader: LoaderFunction = async ({params}): Promise<LoaderData> => {
  return await getPaginatedResults(parseInt(params.page as string));
};

// https://remix.run/api/conventions#meta
export const meta: MetaFunction = () => {
  return {
    title: "Results",
  };
};

export default function MyRoute() {
  const {items, pagination} = useLoaderData<LoaderData>();
  return (
    <div>
      <h1>Results</h1>
      <div>
        {items.map(result => (
          <ResultCard result={result} key={result.id} />
        ))}
      </div>
      <Pagination pagination={pagination} />
    </div>
  );
}
```

and in my `index.tsx` route I re-export from `$page.tsx`

```ts
import MyRoute, {meta} from "~/routes/results/$page";

type LoaderData = {items: Result[], pagination: PaginationInfo}

export const loader: LoaderFunction = async (): Promise<LoaderData> => {
  return await getPaginatedResults(); // I could also pass "1" here
};

export {
  meta
};

export default MyRoute;
```

--

Is there a cleaner approach for this? Something obvious I'm missing? 🙏🏼

**Comment by AlexandrLi** [comment]
Maybe it is a good idea to use URL search params instead of a route?
https://remix.run/docs/en/v1/guides/data-loading#url-search-params

**Reply by machour** [reply]
@AlexandrLi That's an option indeed, but I'd really like to make the parameter part of the route

**Comment by ryanflorence** [comment, marked as answer]
You got it.

"Optional params" in Remix means two routes that call the same loaders and components, but one has a dynamic param, the other is hard coded.

I'd probably move the component, data loading function, and types into a different file and import it into both routes, but it really doesn't matter.

```js
// components/results.jsx
export function getPaginatedResults(page) {
  // ...
}

export function Results({ data }) {
  return (
    <div>{data}</div>
  )
}
```

```js
// routes/results/index.jsx
import { Results, getPaginatedResults } from "~/components/results"

export function loader() {
  return getPaginatedResults(1);
}

export default function Index() {
  return <Results data={useLoaderData()} />
}
```

```js
// routes/results/$page.jsx
import { Results, getPaginatedResults } from "~/components/results"

export function loader({ params ) {
  return getPaginatedResults(parseInt(params.page, 10));
}

export default function Page() {
  return <Results data={useLoaderData()} />
}
```

**Comment by abenhamdine** [comment]
> "Optional params" in Remix means two routes that call the same loaders and components, but one has a dynamic param, the other is hard coded.  

Thx for the hint.  
However I think it's far from being ideal, and doesn't scale well.  
I assume optional parameters should be first reintroduced in react-router, see  https://github.com/remix-run/react-router/issues/8381, before Remix can handle them.  

See also https://github.com/remix-run/remix/issues/1898 and the corresponding PR https://github.com/remix-run/remix/pull/3970, which implements definition of multiple routes for the same component.
